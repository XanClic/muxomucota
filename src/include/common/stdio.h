#ifndef _STDIO_H
#define _STDIO_H

#include <stdarg.h>
#include <stdbool.h>
#include <stddef.h>
#include <vfs.h>


#define EOF (-1)


typedef struct FILE
{
    int fd;
    char buffer[16];
    size_t bufsz;
    bool eof;
} FILE;


int putchar(int c);

int puts(const char *s);

int printf(const char *s, ...);
int fprintf(FILE *fp, const char *s, ...);
int sprintf(char *buf, const char *s, ...);
int snprintf(char *str, size_t n, const char *format, ...);
int vfprintf(FILE *fp, const char *format, va_list args);
int vprintf(const char *format, va_list args);
int vsprintf(char *str, const char *format, va_list args);
int vsnprintf(char *buffer, size_t n, const char *format, va_list argptr);


static inline int filepipe(FILE *fp) { return fp->fd; }

#endif
