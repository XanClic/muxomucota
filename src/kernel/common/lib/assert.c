#include <kassert.h>
#include <panic.h>
#include <stdnoreturn.h>
#include <string.h>


noreturn void assertion_panic(const char *file, int line, const char *function, const char *condition)
{
    size_t buflen = strlen(file) + strlen(function) + strlen(condition) + 64;
    char buffer[buflen];

    int i = 11;
    char linestr[i];
    linestr[--i] = 0;

    while (line != 0)
    {
        linestr[--i] = line % 10 + '0';
        line /= 10;
    }

    strcpy(buffer, "Internal exception: assertion failed\n\n");
    strcat(buffer, file);
    strcat(buffer, ":");
    strcat(buffer, &linestr[i]);
    strcat(buffer, ": ");
    strcat(buffer, function);
    strcat(buffer, "():\n");
    strcat(buffer, condition);


    panic(buffer);
}
