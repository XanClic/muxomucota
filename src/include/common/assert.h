/************************************************************************
 * Copyright (C) 2013 by Hanna Reitz                                    *
 *                                                                      *
 * This file is part of µxoµcota.                                       *
 *                                                                      *
 * µxoµcota  is free  software:  you can  redistribute it and/or modify *
 * it under the terms of the GNU General Public License as published by *
 * the  Free Software Foundation,  either version 3 of the License,  or *
 * (at your option) any later version.                                  *
 *                                                                      *
 * µxoµcota  is  distributed  in the  hope  that  it  will  be  useful, *
 * but  WITHOUT  ANY  WARRANTY;  without even the  implied warranty  of *
 * MERCHANTABILITY  or  FITNESS  FOR  A  PARTICULAR  PURPOSE.  See  the *
 * GNU General Public License for more details.                         *
 *                                                                      *
 * You should have  received a copy of the  GNU  General Public License *
 * along with µxoµcota.  If not, see <http://www.gnu.org/licenses/>.    *
 ************************************************************************/

#ifdef assert
#undef assert
#endif

#ifndef NDEBUG

#include <stdio.h>
#include <stdlib.h>


#ifndef __quote
#define __squote(arg) #arg
#define __quote(arg) __squote(arg)
#endif

#define assert(assertion) do { if (!(assertion)) { printf(__FILE__ ":" __quote(__LINE__) ": "); printf(__func__); printf(": Assertion '" #assertion "' failed.\n"); abort(); } } while (0)

#endif
