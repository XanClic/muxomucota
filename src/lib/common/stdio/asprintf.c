/************************************************************************
 * Copyright (C) 2013 by Hanna Reitz                                    *
 *                                                                      *
 * This file is part of µxoµcota.                                       *
 *                                                                      *
 * µxoµcota  is free  software:  you can  redistribute it and/or modify *
 * it under the terms of the GNU General Public License as published by *
 * the  Free Software Foundation,  either version 3 of the License,  or *
 * (at your option) any later version.                                  *
 *                                                                      *
 * µxoµcota  is  distributed  in the  hope  that  it  will  be  useful, *
 * but  WITHOUT  ANY  WARRANTY;  without even the  implied warranty  of *
 * MERCHANTABILITY  or  FITNESS  FOR  A  PARTICULAR  PURPOSE.  See  the *
 * GNU General Public License for more details.                         *
 *                                                                      *
 * You should have  received a copy of the  GNU  General Public License *
 * along with µxoµcota.  If not, see <http://www.gnu.org/licenses/>.    *
 ************************************************************************/

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>

int asprintf(char **strp, const char *format, ...)
{
    va_list args1, args2;
    int rval;

    va_start(args1, format);
    va_copy(args2, args1);
    rval = vsnprintf(NULL, 0, format, args1);
    va_end(args1);

    *strp = malloc(rval + 1);
    vsnprintf(*strp, rval + 1, format, args2);
    va_end(args2);

    return rval;
}
