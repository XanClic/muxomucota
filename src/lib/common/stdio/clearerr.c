#include <stdio.h>

void clearerr(FILE *stream)
{
    stream->eof = false;
}
