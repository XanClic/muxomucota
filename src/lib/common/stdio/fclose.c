#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <vfs.h>

int fclose(FILE *stream)
{
    int fd = stream->fd;
    free(stream);

    destroy_pipe(fd, 0);

    return 0;
}
