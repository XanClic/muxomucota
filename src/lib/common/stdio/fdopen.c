/************************************************************************
 * Copyright (C) 2013 by Hanna Reitz                                    *
 *                                                                      *
 * This file is part of µxoµcota.                                       *
 *                                                                      *
 * µxoµcota  is free  software:  you can  redistribute it and/or modify *
 * it under the terms of the GNU General Public License as published by *
 * the  Free Software Foundation,  either version 3 of the License,  or *
 * (at your option) any later version.                                  *
 *                                                                      *
 * µxoµcota  is  distributed  in the  hope  that  it  will  be  useful, *
 * but  WITHOUT  ANY  WARRANTY;  without even the  implied warranty  of *
 * MERCHANTABILITY  or  FITNESS  FOR  A  PARTICULAR  PURPOSE.  See  the *
 * GNU General Public License for more details.                         *
 *                                                                      *
 * You should have  received a copy of the  GNU  General Public License *
 * along with µxoµcota.  If not, see <http://www.gnu.org/licenses/>.    *
 ************************************************************************/

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <vfs.h>

FILE *fdopen(int fd, const char *mode)
{
    if ((fd < 0) || (fd >= __MFILE))
    {
        errno = EBADF;
        return NULL;
    }

    FILE *fp = malloc(sizeof(*fp));
    fp->fd = fd;
    fp->bufsz = 0;
    fp->eof = false;

    (void)mode;

    return fp;
}
