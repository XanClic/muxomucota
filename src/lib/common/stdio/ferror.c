#include <stdio.h>

int ferror(FILE *stream)
{
    (void)stream;

    return 0;
}
