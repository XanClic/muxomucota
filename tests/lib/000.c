#include <assert.h>
#include <stdio.h>

/* TEST
   { headers: ['stdlib.h', 'lock.h'],
     files: ['malloc', 'free'].map { |f| "stdlib/#{f}.c" } }
 */

void lock(volatile lock_t *v)
{
    assert(!__sync_val_compare_and_swap(v, 0, 1));
}

void unlock(volatile lock_t *v)
{
    __sync_lock_release(v);
}

void *sbrk(intptr_t sz)
{
    return NULL;
}

void abort(void)
{
    fprintf(stderr, "[abort] Exiting gracefully\n");
    exit(0);
}

int main(void)
{
    void *foo = malloc(8192);
    // will never get here
    printf("%p\n", foo);
    free(foo);
    return 0;
}
