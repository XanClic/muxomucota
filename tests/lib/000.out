[abort] Exiting gracefully
common/stdlib/malloc.c:87: malloc: Assertion 't' failed.
